<a href="https://www.kaggle.com/code/mohamedbakrey/day-7-to-learn-pytorch-linear-regression-1d-pred?scriptVersionId=106349759" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

<h1><center>Linear Regression 1D: Prediction</center></h1>


<h2>Objective</h2><ul><li> How to make the prediction for multiple inputs.</li><li> How to use linear class to build more complex models.</li><li> How to build a custom module.</li></ul> 


<h2>Table of Contents</h2>

<p>In this NoteBook, we will  review how to make a prediction in several different ways by using PyTorch.
<ul>
    <li><a href="https://#Prediction">Prediction</a></li>
    <li><a href="https://#Linear">Class Linear</a></li>
    <li><a href="https://#Cust">Build Custom Modules</a></li>
</ul>

<h2>Preparation</h2>


In [2]:
# These are the libraries will be used for this Notebook.
import torch

<h1 id="Prediction">Prediction</h1>


Let us create the following expressions:


<center>
$b=-1,w=2$ </center> 

<center>
$\hat{y}=-1+2x$ </center> 


First, define the parameters:


In [3]:
# Define w = 2 and b = -1 for y = wx + b

w = torch.tensor(2.0, requires_grad = True)
b = torch.tensor(-1.0, requires_grad = True)

Then, define the function <code>forward(x, w, b)</code> makes the prediction:


In [4]:
# Function forward(x) for prediction
def forward(x):
    yhat=w *x+b
    return yhat

Let's make the following prediction at <i>x = 1</i>


<center> $\hat{y}=-1+2x$</center>


<center> $\hat{y}=-1+2(1)$</center> 



In [5]:
# Predict y = 2x - 1 at x = 1

x=torch.tensor([[0.1]])
yhat=forward(x)
print("The Prediction :", yhat)

The Prediction : tensor([[-0.8000]], grad_fn=<AddBackward0>)


Now, let us try to make the prediction for multiple inputs:
  
  
  
<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2.png" width="500" alt="Linear Regression Multiple Input Samples" />


Let us construct the <code>x</code> tensor first. Check the shape of <code>x</code>.


In [6]:
# Create x Tensor and check the shape of x tensor
x= torch.tensor([[0.1],[2.0]])
print("The shape of x: ",x.shape)

The shape of x:  torch.Size([2, 1])


Now we can make the prediction

In [7]:
# Make the prediction of y = 2x - 1 at x = [1, 2]
yhat= forward(x)
print("The prediction:", yhat)

The prediction: tensor([[-0.8000],
        [ 3.0000]], grad_fn=<AddBackward0>)


The result is the same as what it is in the image above.


<h1 id="Linear">Class Linear</h1>

The linear class can be used to make a prediction. We can also use the linear class to build more complex models. Let's import the module:


In [8]:
# Import Class Linear
from torch.nn import Linear

Set the random seed because the parameters are randomly initialized:


In [13]:
# Set random seed

torch.manual_seed(1)

Now we can create the linear object by using the constructor. The parameters are randomly created. Let us print out to see what <i>w</i> and <i>b</i>. The parameters of an <code>torch.nn.Module</code> model are contained in the model’s parameters accessed with <code>lr.parameters()</code>:


In [16]:
# Create Linear Regression Model, and print out the parameters
lr= Linear(in_features=1, out_features=1,bias=True)
print("Parameter w and b :", list(lr.parameters()))

Parameter w and b : [Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]


This is equivalent to the following expression:

<center>$b=-0.44, w=0.5153$</center>

<center>$\hat{y}=-0.44+0.5153x$</center>


A method  <code>state_dict()</code> Returns a Python dictionary object corresponding to the layers of each parameter  tensor.


In [20]:
print("Python dictionary: ",lr.state_dict())
print("********************************************")
print("keys: ",lr.state_dict().keys())
print("********************************************")
print("values: ",lr.state_dict().values())

Python dictionary:  OrderedDict([('weight', tensor([[0.5153]])), ('bias', tensor([-0.4414]))])
********************************************
keys:  odict_keys(['weight', 'bias'])
********************************************
values:  odict_values([tensor([[0.5153]]), tensor([-0.4414])])


The keys correspond to the name of the attributes and the values correspond to the parameter value.


In [19]:
print("weight:",lr.weight)
print("********************************************")
print("bias:",lr.bias)

weight: Parameter containing:
tensor([[0.5153]], requires_grad=True)
********************************************
bias: Parameter containing:
tensor([-0.4414], requires_grad=True)


Now let us make a single prediction at <i>x = \[\[1.0]]</i>.


In [21]:
# Make the prediction at x = [[1.0]]

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739]], grad_fn=<AddmmBackward0>)


Similarly, you can make multiple predictions:


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2vector_function.png" width="500" alt="Linear Class Sample with Multiple Inputs" />


Use model <code>lr(x)</code> to predict the result.


In [22]:
# Create the prediction using linear model

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739],
        [0.5891]], grad_fn=<AddmmBackward0>)


<h1 id="Cust">Build Custom Modules</h1>


Now, let's build a custom module. We can make more complex models by using this method later on.


In [23]:
# Library for this section

from torch import nn

Now, let us define the class:


In [24]:
# Customize Linear Regression Class

class LR(nn.Module):
    
    # Constructor
    def __init__(self, input_size, output_size):
        
        # Inherit from parent
        super(LR, self).__init__()
        self.linear = nn.Linear(input_size, output_size)
    
    # Prediction function
    def forward(self, x):
        out = self.linear(x)
        return out

Create an object by using the constructor. Print out the parameters we get and the model.


In [25]:
# Create the linear regression model. Print out the parameters.

lr = LR(1, 1)
print("The parameters: ", list(lr.parameters()))
print("**************************************************************")
print("Linear model: ", lr.linear)

The parameters:  [Parameter containing:
tensor([[-0.1939]], requires_grad=True), Parameter containing:
tensor([0.4694], requires_grad=True)]
**************************************************************
Linear model:  Linear(in_features=1, out_features=1, bias=True)


Let us try to make a prediction of a single input sample.


In [26]:
# Try our customize linear regression model with single input

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755]], grad_fn=<AddmmBackward0>)


Now, let us try another example with multiple samples.


In [27]:
# Try our customize linear regression model with multiple input

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755],
        [0.0816]], grad_fn=<AddmmBackward0>)


In [29]:
print("Python dictionary: ", lr.state_dict())
print("**************************************************************")
print("keys: ",lr.state_dict().keys())
print("**************************************************************")
print("values: ",lr.state_dict().values())


Python dictionary:  OrderedDict([('linear.weight', tensor([[-0.1939]])), ('linear.bias', tensor([0.4694]))])
**************************************************************
keys:  odict_keys(['linear.weight', 'linear.bias'])
**************************************************************
values:  odict_values([tensor([[-0.1939]]), tensor([0.4694])])


# Summary...

In this work, we work on explaining the basics that are used in deep learning using PyTroch.

You can find the first day  [here](https://www.kaggle.com/code/mohamedbakrey/day-1-to-learn-pytorch-torch-tensors-in-1d)

You can find the second day  [here](https://www.kaggle.com/code/mohamedbakrey/day-2-to-learn-pytorch-two-dimensional-tensors)

You can find the Thrid day  [here](https://www.kaggle.com/code/mohamedbakrey/day-3-to-learn-pytorch-differentiation-in-pytorc)

You can find the Fourth day [here](https://www.kaggle.com/code/mohamedbakrey/day-4-to-learn-pytorch-simple-dataset)

You can find the Five day [here](https://www.kaggle.com/code/mohamedbakrey/day-5-to-learn-pytorch-image-datasets-and-transf)

You can find the sex day [here](https://www.kaggle.com/code/mohamedbakrey/day-6-to-learn-pytorch-prebuilt-d-t)